### DISCUSSION TITLE
Utils.auto_refreshing - what's the difference between self.refresh_data_bindings and the forms `refreshing_data_bindings` event

### DISCUSSION THREAD

**Discussion by rhurlbatt** [anvilistas/anvil-extras#249] [Q&A] [answered]
I see in this code that it relies on `self.refresh_data_bindings`. Does this function also call the form's `refreshing_data_bindings` event indirectly?

My thought was to use this as a way of running a form input validation check and enabling or disabling a save button whenever data is entered. This way if you are editing an item, you can also check if it has changed. For example, when ever `auto_refreshing` runs `self.refresh_data_bindings` it should also run the form's `refreshing_data_bindings` event. Therefor I could write my validation code in the `form_refreshing_data_bindings()` function which is called the event and have `auto_refresh `run it any time the tiem changed.

I originally thought that the forms `refreshing_data_bindings` event (not `self.refresh_data_bindings`) was called any time the item changed but it does not seem to be the case.

**Comment by s-cork** [comment, marked as answer]
The event triggers any time:

The form's item is changed like:
`self.item = self.item`
Or
You invoke `refresh_data_bindings`

`refresh_data_bindings` is equivalent to 
`self.item = self.item`. 


Side note - the anvil-labs atomic module is like a super charged version of `auto_refreshing`. 


Edit:
If the above doesn't seem to be the case - clone link?

**Reply by rhurlbatt** [reply]
It looks like a bug - see clone link. [see clone link.](https://anvil.works/build#clone:US6ZDGRI4OLPJME2=XGWBI7CHHYO77HZ2A77BJBLR)

When the textbox is filled, you can verify `self.item` has changed but the `form_refreshing_data_bindings()` event is not called

**Reply by s-cork** [reply]
Right you are. 

If you do self.item = self.item in the init method it should work. 

Or pass `self.init_components(item={})`

The bug happens if the form is never invoked with an item.
ie the `**properties` doesn't include an item. 

I think to fix the bug we'd need to override init_components in a similar way to how we override the item property. 

something like 

```python
def mk_init_components(cls):
    super_init = super(cls, cls).init_components
    def init_components(self, item=None, **props):
        if item is None:
            item = {}
        super_init(self, item=item, **props)
    return init_components
```

**Reply by rhurlbatt** [reply]
Using your code above, and hacking together the rest of it - this works as expected but I am not sure if I have missed anything.  ([same clone link](https://anvil.works/build#clone:US6ZDGRI4OLPJME2=XGWBI7CHHYO77HZ2A77BJBLR))
```python
def _mk_init_components(cls):
    """Generate a method to override a form's 'init_components' method"""
    super_init = super(cls, cls).init_components
    def init_components(self, item=None, **props):
        if item is None:
            item = {}
        super_init(self, item=item, **props)
    return init_components

  
@cache
def auto_refreshing(cls):
    """A decorator for a form class to refresh data bindings automatically"""
    # Override the form class' item attribute
    cls.item = _item_override(cls)
    
     # Override the form class' init method
    cls.init_components = _mk_init_components(cls)
    
    return cls
```
So for my own reason and understanding, is this how the code works:

 When the `@auto_refreshing` decorator is applied, it over writes the superclass' `init_components()` with with a new `init_components()` that simply replaces the item if it is `None` 
```python
def init_components(self, item=None, **props):
        if item is None:
            item = {}
        super_init(self, item=item, **props)
```
Which basically changes `self.item` to an empty dict if the `self.item is None`.

The `super()` method is used to inherit the form's `init_components()` method without recreating it (ie the `auto_refreshing` module doesn't know how `init_components()` works but because we are inheriting it doesn't need to)?

Do I understand this correctly?

**Reply by s-cork** [reply]
Yeah it's basically a convoluted way of doing the following.

```
class A(B):
  def __init__(self):
    super().__init__()
    self. a = 1

# equivalent to
class A(B):
  def __init__(self):
    super(A, self).__init__()
    self. a = 1
```

The super object takes the parameters and walks the base classes to finds the next base class that has the method you're after.
In our example the "next in line" for `init_components` is the FormTemplate's `init_components` method.

We could even clean up the calls by creating the super object in the `auto_refreshing` decorator and passing this to the override helpers instead of the cls.

Looking forward to the pr 😉

**Reply by rhurlbatt** [reply]
Close but no cigar - if a component is bound to a dict within `self.item`, it still does not call the ` form_refreshing_data_bindings()` event. 
for example: `self.text_box_1.text` is bound to `self.item['data_dict']['text']`

If the component bound to a key in the base dict this works as expected.

See [same clone link  ](https://anvil.works/build#clone:US6ZDGRI4OLPJME2=XGWBI7CHHYO77HZ2A77BJBLR)
I am not sure how to get around this or if it is even possible

🤔

**Reply by s-cork** [reply]
The clone link works as expected?
The refresh databindings is triggered on the lost focus or pressed enter event - as you'd expect. 

Yeah nested dictionaries aren't implemented.

**Reply by rhurlbatt** [reply]
> The clone link works as expected?

Yes but not with nested dicts - was not sure if this was the case originally.

I will create a PR for code and docs accordingly.

Thanks for your help Stu (or more correctly: Thanks for fixing this issue Stu)

**Reply by s-cork** [reply]
Good stuff. 

I'm gonna keep plugging the anvil-labs atomic module. 

Nested dicts will just work there.
